proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+- if it is of shape `InfoTablePtrOf# '〈d〉`, where `〈d〉` is a data constructor, it returns the info table pointer for data constructor `〈d〉`
+- otherwise, it raises a compile-time error (I've yet to find how to do that in practice)
+
+Defining `reifyInfoTablePtr#` requires support for kind-polymorphic type variables in primop generation. I include the following changes in the proposal, following what has been done for levity-polymorphic type variables:
+
+```haskell
+kind1TyVar, kind2TyVar :: TyVar
+(kind1TyVar : kind2TyVar : _)
+  = mkTemplateKindVars (repeat liftedTypeKind)
+
+kind1TyVarInf, kind2TyVarInf :: TyVarBinder
+kind1TyVarInf = mkTyVarBinder Inferred kind1TyVar
+kind2TyVarInf = mkTyVarBinder Inferred kind2TyVar
+
+kind1Ty, kind2Ty :: Type
+kind1Ty = mkTyVarTy kind1TyVar
+kind2Ty = mkTyVarTy kind2TyVar
+
+kindPolyAlphaTyVar, kindPolyBetaTyVar :: TyVar
+[kindPolyAlphaTyVar, kindPolyBetaTyVar] =
+  mkTemplateTyVars
+    [ kind1Ty
+    , kind2Ty]
+
+kindPolyAlphaTyVarSpec, kindPolyBetaTyVarSpec :: TyVarBinder
+kindPolyAlphaTyVarSpec = mkTyVarBinder Specified kindPolyAlphaTyVar
+kindPolyBetaTyVarSpec  = mkTyVarBinder Specified kindPolyBetaTyVar
+
+kindPolyAlphaTy, kindPolyBetaTy :: Type
+kindPolyAlphaTy = mkTyVarTy kindPolyAlphaTyVar
+kindPolyBetaTy  = mkTyVarTy kindPolyBetaTyVar
+```
+
+### Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family
+
+```haskell
+type family LiftedCtorToSymbol (liftedCtor :: k) :: Symbol
+```
+
+When making an API around `compactAddHollow#`, we need to know the number and type of the fields that the constructor has. That information can be easily retrieved with `GHC.Generics`. However, the type-level structure returned by `GHC.Generics` distinguishes constructors of a given type by their name, as a `Symbol`.
+
+I introduce a built-in type family `LiftedCtorToSymbol` whose in charge of translating a data constructor lifted as a type into the corresponding symbol. That way a library can leverage `GHC.Generics` to allow direct and safe structure building in compact regions for most user-defined data types.
+
+For example:
+
+```haskell
+LiftedCtorToSymbol 'Left ~ "Left"
+```
+
+More generally, forall data constructor `〈d〉`,
+
+```haskell
+LiftedCtorToSpecCtor '〈d〉 ~ "〈d〉"
+```
+
+## Proposed Library Change Specification
+
+- `compactAddHollow#`, `reifyInfoTablePtr#`, and `InfoTablePtrOf#` are exposed in `GHC.Prim`
+- `LiftedCtorToSymbol` is exposed in `GHC.TypeLits`
+
+## Examples
+
+### Primitive usage
+
+#### Allocate a hollow `Left` constructor in region `compactRegion`
+
+```haskell
+let leftInfoPtr :: InfoTablePtrOf# 'Left = reifyInfoTablePtr# (# #)
+ in compactAddHollow# compactRegion (unsafeCoerceAddr leftInfoPtr) state
+```
+
+Yet unspecified fields can later be assigned with existing primop `writeAddrOffAddr#` (we know the offsets of the fields thanks to the return type of `compactAddHollow#`).
```

Comment:
Remove the parenthetical remark
```suggestion
Yet unspecified fields can later be assigned with existing primop `writeAddrOffAddr#`.
```

Replace it by extending the example so that it uses the offsets.

Reply:
I extended the example as requested. It's fairly more complex now, I hope it's still readable.